File: USACOContest/2015Dec/bcount/bcount.cpp
/*
ID: ulysses4
LANG: C++
PROG: bcount
*/
#include <cstdio>
#include <cstring>
char PROG[] = "bcount";

int n, q, sum[100001][4];
void actualMain() {
	scanf("%d%d", &n, &q);
	for (int i = 1, breed; i <= n; i++) {
		scanf("%d", &breed);
		for (int b = 1; b <= 3; b++)
			sum[i][b] = sum[i - 1][b] + (b == breed);
	}
	for (int a, b; q; q--) {
		scanf("%d%d", &a, &b);
		printf("%d %d %d\n",
				sum[b][1] - sum[a - 1][1],
				sum[b][2] - sum[a - 1][2],
				sum[b][3] - sum[a - 1][3]);
	}
}

int main() {
	int size = sizeof(PROG);
	char in[size + 3], out[size + 4];
	freopen(strcat(strcpy(in, PROG), ".in"), "r", stdin);
	freopen(strcat(strcpy(out, PROG), ".out"), "w", stdout);
	actualMain();
	fclose(stdin);
	fclose(stdout);
	return 0;
}

File: USACOContest/2015Dec/bcount/bcount.java
import java.io.*;
import java.util.*;
public class bcount {
	private static final String PROG = "bcount";
	
	private static int q;
	private static int[][] sum;
	private static void init() {
		int n = scanner.nextInt();
		q = scanner.nextInt();
		sum = new int[n + 1][4];
		for (int i = 1; i <= n; i++) {
			int breed = scanner.nextInt();
			for (int b = 1; b <= 3; b++)
				sum[i][b] = sum[i - 1][b] + (b == breed ? 1 : 0);
		}
	}
	private static void output(int a, int b, int breed) {
		out.print(sum[b][breed] - sum[a - 1][breed]);
	}
	private static void output() {
		while (q-- > 0) {
			int a = scanner.nextInt();
			int b = scanner.nextInt();
			output(a, b, 1);
			out.print(' ');
			output(a, b, 2);
			out.print(' ');
			output(a, b, 3);
			out.println();
		}
	}
	private static void actualMain() {
		init();
		output();
	}
	
	private static InputStream in;
	private static PrintStream out;
	private static Scanner scanner;
	public static void main(String[] args) throws IOException {
		in = new FileInputStream(PROG + ".in");
		out = new PrintStream(PROG + ".out");
		scanner = new Scanner(in);
		actualMain();
		in.close();
		out.close();
		scanner.close();
	}
}
